Update docstrings, add upgrading guide (#231)
bytewax/bytewax: docs/articles/reference/upgrading-to-v16.md

## Upgrading to v0.16

Bytewax v0.16 introduces major changes to the way that custom Bytewax inputs are created, as well as improvements to and windowing and execution. In this article, we'll cover some of the updates we've made to our API to make upgrading to v0.16 easier.

> ## Input changes

Reviewer: Should we also call out that `bytewax.inputs.KafkaInputConfig` has been moved to `bytewax.connectors.kafka.KafkaInput`?

In v0.16, we have restructured input to be based around partitions in order to support rescaling stateful dataflows in the future. We have also dropped the `Config` suffix to our input classes. As an example, `KafkaInputConfig` has been renamed to `KafkaInput`.

`ManualInputConfig` is now replaced by two base superclasses that can be subclassed to create custom input sources. They are `DynamicInput` and `PartitionedInput`.

### `DynamicInput`

`DynamicInput` sources are input sources that support reading distinct items from any number of workers concurrently.

> Dynamic inputs must be able to supply disjoint data for each worker that is started. If multiple workers read the same data, the Dataflow will process those as duplicate items.

Reviewer: This is true for all inputs, not just dynamic ones, and it something that is touched upon in the input docstrings. I'd just delete this line.

> `DynamicInput` sources do not support storing resume state.

Reviewer: ```suggestion
`DynamicInput` sources do not support resume state and thus generally do not provide delivery guarantees better than at-most-once.
```

### `PartitionedInput`

`PartitionedInput` sources are input sources that manage a fixed number of independent partitions, like Kafka.

> `PartitionedInput` sources maintain the state of each source and re-build that state during recovery.

Reviewer: ```suggestion
`PartitionedInput` sources can keep internal state on the current position of each partition and if they can rewind and resume reading from those previous on resume, can provide delivery guarantees of at-least-once or better.
```

## Capture -> Output

> Outputs in v0.16 are now stateful, and are now configured with a name for the step:

Reviewer: ```suggestion
The `capture` operator has been renamed to `output` in v0.16 and is now stateful, so requires a step ID:
```

In v0.15 and before:

``` python doctest:SKIP
flow.capture(StdOutputConfig())
```

In v0.16+:

``` python
from bytewax.dataflow import Dataflow
from bytewax.connectors.stdio import StdOutput

flow = Dataflow()
flow.output("out", StdOutput())
```

> 

Reviewer: ```suggestion

`ManualOutputConfig` is now replaced by two base superclasses that can be subclassed to create custom output sinks. They are `DynamicOutput` and `PartitionedOutput`.

```

> ## New entrypoint

Reviewer: Do we want to call out at the top level that Kafka recovery is being deprecated? Should we also add that to `CHANGELOG`?

In Bytewax v0.16, the way that Dataflows are run has been simplified, and most execution functions have been removed.

> Similar to other Python frameworks like Flask and FastAPI, Dataflows can now be run using a Datfaflow import string in the <module_name>:<dataflow_variable_name_or_factory_function> format. As an example, given a file named `dataflow.py` with the following contents:

Reviewer: I forgot, does it default to looking for a `flow` variable? Can you leave off the `:flow`?

``` python
from bytewax.dataflow import Dataflow
from bytewax.testing import TestingInput
from bytewax.connectors.stdio import StdOutput

flow = Dataflow()
flow.input("in", TestingInput(range(3)))
flow.output("out", StdOutput())
```

> The Dataflow can be run with the following invocation:

Reviewer: ```suggestion
Since a Python file `dataflow.py` defines a module `dataflow`, the Dataflow can be run with the following invocation:
```

        yield state, line

def lower(line):
    return line.lower()

def tokenize(line):
    return re.findall(r'[^\s!,.?":;0-9]+', line)

def initial_count(word):
    return word, 1

def add(count1, count2):
    return count1 + count2

clock_config = SystemClockConfig()
window_config = TumblingWindowConfig(length=timedelta(seconds=5))

flow = Dataflow()
flow.input("input", ManualInputConfig(input_builder))
flow.map(lower)
flow.flat_map(tokenize)
flow.map(initial_count)
flow.reduce_window("sum", clock_config, window_config, add)
flow.capture(StdOutputConfig())

run_main(flow)
```

To port the example to the `0.16` version we need to make a few changes.

### Imports

Let's start with the existing imports:

```python doctest:SKIP
from bytewax.inputs import ManualInputConfig
from bytewax.outputs import StdOutputConfig
from bytewax.execution import run_main
```

Becomes:
```python doctest:SKIP
from bytewax.connectors.files import FileInput
from bytewax.connectors.stdio import StdOutput
```

We removed `run_main` as it is now only used for unit testing dataflows. Bytewax now has a built-in FileInput connector, which uses the `PartitionedInput` connector superclass.

> Since we are using a built-in connector to read from this file, we can delete our `file_input` function above.

Reviewer: There is no `file_input` function above.

from bytewax.dataflow import Dataflow
from bytewax.connectors.files import FileInput
from bytewax.connectors.stdio import StdOutput
from bytewax.window import SystemClockConfig, TumblingWindow

def lower(line):
    return line.lower()

def tokenize(line):
    return re.findall(r'[^\s!,.?":;0-9]+', line)

def initial_count(word):
    return word, 1

def add(count1, count2):
    return count1 + count2

clock_config = SystemClockConfig()
window_config = TumblingWindow(
    length=timedelta(seconds=5), align_to=datetime(2023, 1, 1, tzinfo=timezone.utc)
)

flow = Dataflow()
flow.input("inp", FileInput("examples/sample_data/wordcount.txt"))
flow.map(lower)
flow.flat_map(tokenize)
flow.map(initial_count)
flow.reduce_window("sum", clock_config, window_config, add)
flow.output("out", StdOutput())
```

Running our completed Dataflow looks like this, as we've named our Dataflow variable `flow` in a file named `dataflow`:

``` bash
python -m bytewax.run dataflow:flow
('whether', 1)
("'tis", 1)
('of', 2)
('opposing', 1)
```

We can even run our sample Dataflow on multiple workers to process the file in parallel:

In one terminal, run:

``` bash

> > python -m bytewax.run dataflow:flow -i0 -a "localhost:2101;localhost:2102"

Reviewer: Is this better demonstrated via the "spawn cluster" `-p` flag, rather than starting two workers by hand?

## Upgrading to v0.16

Bytewax v0.16 introduces major changes to the way that custom Bytewax inputs are created, as well as improvements to and windowing and execution. In this article, we'll cover some of the updates we've made to our API to make upgrading to v0.16 easier.

## Input changes

In v0.16, we have restructured input to be based around partitions in order to support rescaling stateful dataflows in the future. We have also dropped the `Config` suffix to our input classes. As an example, `KafkaInputConfig` has been renamed to `KafkaInput` and has been moved to `bytewax.connectors.kafka.KafkaInput`.

`ManualInputConfig` is now replaced by two base superclasses that can be subclassed to create custom input sources. They are `DynamicInput` and `PartitionedInput`.

### `DynamicInput`

`DynamicInput` sources are input sources that support reading distinct items from any number of workers concurrently.

`DynamicInput` sources do not support resume state and thus generally do not provide delivery guarantees better than at-most-once.

### `PartitionedInput`

> `PartitionedInput` sources can keep internal state on the current position of each partition. If a partition can "rewind" and resume reading from a previous position, they can provide delivery guarantees of at-least-once or better.

Reviewer 2: by `or better`, do we mean exactly once with some caveats? Maybe we could be more specific here
